discovery_engine_core/bindings/src/types/duration.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! FFI functions for handling `Duration` fields.
+
+use std::time::Duration;
+
+/// Initializes a `Duration` field at given place.
+///
+/// # Safety
+///
+/// It must be valid to write a [`Duration`] to given pointer.
+#[no_mangle]
+pub unsafe extern "C" fn init_duration_at(place: *mut Duration, seconds: u64, nanos: u32) {
+    unsafe {
+        place.write(Duration::new(seconds, nanos));
```

Comment:
related to

> > Can't we encoded duration as a number since we are interested only in seconds and avoid pointers?
>
> We can't avoid pointer as we write to a field of a struct. Even with just a u32 there would be an pointer.
>
> We also decided to use Duration in rust, not u32/u64.

if we only expect seconds granularity from dart, we could simplify the ffi api a bit and still get the benefits of the `Duration` api internally in rust

```suggestion
pub unsafe extern "C" fn init_duration_at(place: *mut Duration, seconds: u64) {
    unsafe {
        place.write(Duration::from_secs(seconds));
```

then we could also remove `get_duration_nanos()`.

Reply:
yes, I considered that, but I preferred to have a dart Duration to rust Duration type we write once and then can use anywhere we need a duration without touching it again.

It also means that we will have an leaky abstraction which seem to do one thing (api for passing a dart/rust Duration through the ffi) but does another thing (API for truncating duration to seconds precision and passing them from/to dart).